src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs

```diff
@@ -15,54 +15,79 @@ internal partial class LocalRewriter
-        /// then lowering the decision tree to a sequence of bound statements. We inherit <see cref="DecisionTreeBuilder"/>
-        /// for the machinery to build the decision tree, and then use
-        /// <see cref="PatternSwitchLocalRewriter.LowerDecisionTree(BoundExpression, DecisionTree)"/>
-        /// recursively to produce bound nodes that implement it.
-        /// </summary>
-        private class PatternSwitchLocalRewriter : DecisionTreeBuilder
+        private class PatternSwitchLocalRewriter : BasePatternSwitchLocalRewriter
         {
-            private readonly LocalRewriter _localRewriter;
-            private readonly HashSet<LocalSymbol> _declaredTempSet = new HashSet<LocalSymbol>();
-            private readonly ArrayBuilder<LocalSymbol> _declaredTemps = ArrayBuilder<LocalSymbol>.GetInstance();
-            private readonly SyntheticBoundNodeFactory _factory;
+            public static BoundNode Rewrite(LocalRewriter localRewriter, BoundPatternSwitchStatement node)
+            {
+                var rewriter = new PatternSwitchLocalRewriter(node, localRewriter);
+                BoundStatement result = rewriter.LowerPatternSwitchStatement(node);
+                rewriter.Free();
+                return result;
+            }
 
             /// <summary>
-            /// Map from switch section's syntax to the lowered code for the section.
+            /// Map from switch section's syntax to the lowered code for the section. The code for a section
+            /// includes the code to assign to the pattern variables and evaluate the when clause. Since a
+            /// when clause can yield a false value, it can jump back to a label in the lowered decision dag.
             /// </summary>
-            private readonly Dictionary<SyntaxNode, ArrayBuilder<BoundStatement>> _switchSections = new Dictionary<SyntaxNode, ArrayBuilder<BoundStatement>>();
+            private Dictionary<SyntaxNode, ArrayBuilder<BoundStatement>> _switchSections => base._switchArms;
 
-            private ArrayBuilder<BoundStatement> _loweredDecisionTree = ArrayBuilder<BoundStatement>.GetInstance();
+            /// <summary>
+            /// A map from section syntax to the first label in that section.
+            /// </summary>
+            private Dictionary<SyntaxNode, LabelSymbol> _sectionLabels = PooledDictionary<SyntaxNode, LabelSymbol>.GetInstance();
 
-            private PatternSwitchLocalRewriter(LocalRewriter localRewriter, BoundPatternSwitchStatement node)
-                : base(localRewriter._factory.CurrentMethod, (SwitchStatementSyntax)node.Syntax, localRewriter._factory.Compilation.Conversions)
+            /// <summary>
+            /// We revise the returned label for a decision so that all decisions in the same switch section are given the same label.
+            /// This permits all the decisions in a
+            /// section to share the label, which enables the switch emitter to produce better code.
```

Comment:
Seems redundant given the previous sentence.

Reply:
Recorded in [https://github.com/dotnet/roslyn/issues/25663](https://github.com/dotnet/roslyn/issues/25663)

---
In reply to: [176308611](https://github.com/dotnet/roslyn/pull/25544#discussion_r176308611) [](ancestors = 176308611)